✂️ refactor: MCP UI Separation for Agents

dustinhealy opened the pull request:
## Summary

This pull request separates MCP servers from the Tools & Actions UI in the Agent Builder into their own MCP Servers list with their own Dialog for adding servers (and configuring custom user variables) with the new 'Add MCP Server Tools' button. It also allows MCP servers that have been previously saved to an agent to be listed and reinitialized even when they are not connected or configured in the YAML (connection state can now be seen in the Agent Builder via the addition of connection status indicators for installed MCP servers in the MCP Servers list). Since disconnected servers can now be added through the new MCPToolSelectDialog, servers will now attempt to reinitialize before they are added to an agent if they are not currently connected. 

**Enhancements to MCP Tools Handling and UI:**

* Refactored the grouping logic for agent tools and MCP tools in `AgentPanelContext`, introducing `groupedMCPTools` for clearer separation and easier access to MCP-related tools. (`client/src/Providers/AgentPanelContext.tsx` [[1]](diffhunk://#diff-20b4634b5101343f5db4fc0d772f14a0b06f178b36a2554366eaa69b01c83bf6L36-R89) [[2]](diffhunk://#diff-20b4634b5101343f5db4fc0d772f14a0b06f178b36a2554366eaa69b01c83bf6R109); `client/src/common/types.ts` [[3]](diffhunk://#diff-72854f32baf56c716453441e1ae7430c9310f31c1ac9a3b46851172103e09820R223)
* Updated the agent configuration panel to display MCP tools in a dedicated section, now showing disconnected MCP servers that were previously added to the agent, and reflecting server connection status. (`client/src/components/SidePanel/Agents/AgentConfig.tsx` [[1]](diffhunk://#diff-1aa1df4cfc3096810fbd224b4b1e3278d7decd8576e543d1c8634dd25986dd99R49-R63) [[2]](diffhunk://#diff-1aa1df4cfc3096810fbd224b4b1e3278d7decd8576e543d1c8634dd25986dd99L388-R524)
* Added a new `UninitializedMCPTool` component to represent MCP servers that are not yet initialized, which do not show a tool-dropdown but can be clicked in order to reinitialize the server, and integrated it into the agent configuration panel. (`client/src/components/SidePanel/Agents/UninitializedMCPTool.tsx` [[1]](diffhunk://#diff-ec4d0733fc3ffcdca2b2a66f6c4a0ef7d4551056c920509c1bf68ef0a57853c2R1-R165); `client/src/components/SidePanel/Agents/AgentConfig.tsx` [[2]](diffhunk://#diff-1aa1df4cfc3096810fbd224b4b1e3278d7decd8576e543d1c8634dd25986dd99R35)
* Split selection dialogs for both standard tools and MCP tools, allowing users to add MCP server tools via a dedicated dialog that allows for custom user variable configuration. (`client/src/components/SidePanel/Agents/AgentConfig.tsx`, `client/src/components/Tools/MCPToolSelectDialog.tsx`) [[1]](diffhunk://#diff-1aa1df4cfc3096810fbd224b4b1e3278d7decd8576e543d1c8634dd25986dd99R616-R620) [[2]](diffhunk://#diff-1aa1df4cfc3096810fbd224b4b1e3278d7decd8576e543d1c8634dd25986dd99L15-R22)



## Change Type

- [x] New feature (non-breaking change which adds functionality)

## Testing

Manual Verification of Functionality in MCPSelect
Confirmation of ability to install / uninstall tools and MCP servers and that they persist when switching agents
Confirmation of proper filtering in new MCPToolSelectDialog and MCP Server list

## Screenshots

<div align=center>

<img width="338" height="171" alt="Screenshot 2025-08-25 at 11 15 44 PM" src="https://github.com/user-attachments/assets/b416d455-202d-4abc-9f32-5da23b6882d7" />
<p><em> Connection Status Indicators now present in Agent Builder MCP server list </em></p>

---

https://github.com/user-attachments/assets/2bb76674-ad75-412e-ad84-a67fc365d07d
<p><em> Disconnected MCPs now displayed and initializable from the MCP server list</em></p>

---

<img width="333" height="949" alt="Screenshot 2025-08-27 at 5 45 08 PM" src="https://github.com/user-attachments/assets/19f69656-6775-4677-936c-572b86654a55" />
<p><em> Previously saved MCPs now displayed and removable from the MCP server list even when not configured in YAML anymore</em></p>

---

https://github.com/user-attachments/assets/62f6e837-8b8e-4b7d-b042-a537ff6703e8
<p><em> Custom User Variables now configurable from the MCP Selection Dialog when adding a new server to an agent</em></p>

---

<img width="1711" height="949" alt="Screenshot 2025-08-25 at 11 18 30 PM" src="https://github.com/user-attachments/assets/27cb1f57-9c61-464a-99b0-f6401ceff74e" />
<p><em>Disconnected MCP servers are now displayed in the MCP Selection Dialog and will attempt to initialize before being added to an agent</em></p>

</div>

## Checklist

- [x] My code adheres to this project's style guidelines
- [x] I have performed a self-review of my own code
- [x] My changes do not introduce new warnings
- [x] Local unit tests pass with my changes

dustinhealy committed:
refactor: MCP UI Separation for Agents (Dustin WIP)

feat: separate MCPs into their own lists away from tools + actions and add the status indicator functionality from chat to their dropdown ui

fix: spotify mcp was not persisting on agent creation

feat: show disconnected saved servers and their tools in agent mcp list in created agents

fix: select-all regression fixed (caused by deleting tools we were drawing from for rendering list)

fix: dont show all mcps, only those installed in agent in list

feat: separate ToolSelectDialog for MCPServerTools

fix: uninitialized mcp servers not showing as added in toolselectdialog

refactor: reduce looping in AgentPanelContext for categorizing groups and mcps

refactor: split ToolSelectDialog and MCPToolSelectDialog functionality (still needs customization for custom user vars)

chore: address ESLint comments

chore: address ESLint comments

feat: one-click initialization on MCP servers in agent builder

fix: stop propagation triggering reinit on caret click

refactor: split uninitialized MCPs component from initialized MCPs

feat: new mcp tool select dialog ui with custom user vars

feat: show initialization state for CUV configurable MCPs too

chore: remove unused localization string

fix: deselecting all tools caused a re-render

fix: remove subtools so removal from MCPToolSelectDialog works more consistently

feat: added servers have all tools enabled by default

feat: mcp server list now alphabetical to prevent annoying ui behavior of servers jumping around depending on tool selection

fix: filter out placeholder group mcp tools from any actual tool calls / definitions

feat: indicator now takes you to config dialog for uninitialized servers

feat: show previously configured mcp servers that are now missing from the yaml

feat: select all enabled by default on first add to mcp server list

chore: address ESLint comments

danny-avila commented on client/src/components/SidePanel/Agents/AgentConfig.tsx:
```diff
@@ -187,6 +197,14 @@ export default function AgentConfig({ createMutation }: Pick<AgentPanelProps, 'c
     }
   });
 
+  Object.entries(allMCPTools ?? {}).forEach(([toolId, toolObj]) => {
```
Something strikes me as off both here and in the previous block, since we are already grouping/looping through available servers/tools in the agent panel context. There's an anti-pattern here to keep transforming the same set of data via loops.

**Specific issues:**
1. **Redundant processing**: `AgentPanelContext` already groups MCP tools into `groupedMCPTools` with a proper data structure, but this component re-processes the same data multiple times with `Object.entries(allMCPTools ?? {}).forEach()` and additional transformations.

2. **Performance overhead**: The IIFE (lines 406-590) contains heavy computation that runs on every render without memoization - multiple `forEach`, `filter`, and `map` operations on the same dataset for different scenarios (configured/unconfigured/disconnected servers).

3. **Separation of concerns**: Complex data processing logic is mixed with rendering logic instead of being handled in the context provider or memoized.

danny-avila committed:
refactor: MCP UI Separation for Agents (Danny WIP)

chore: remove use of `{serverName}_mcp_{serverName}`

chore: import order

WIP: separate component concerns

refactor: streamline agent mcp tools

refactor: unify MCP server handling and improve tool visibility logic, remove unnecessary normalization or sorting, remove nesting button, make variable names clear

refactor: rename mcpServerIds to mcpServerNames for clarity and consistency across components

refactor: remove groupedMCPTools and toolToServerMap, streamline MCP server handling in context and components to effectively utilize mcpServersMap

refactor: optimize tool selection logic by replacing array includes with Set for improved performance

chore: add error logging for failed auth URL parsing in ToolCall component

refactor: enhance MCP tool handling by improving server name management and updating UI elements for better clarity

danny-avila commented on client/src/components/SidePanel/Agents/AgentConfig.tsx:
```diff
@@ -42,16 +47,21 @@ export default function AgentConfig({ createMutation }: Pick<AgentPanelProps, 'c
+    groupedMCPTools: allMCPTools,
   } = useAgentPanelContext();
 
+  const { connectionStatus } = useMCPServerManager();
```
This highlights that `useMCPServerManager` is likely doing too much. For the "MCP Section" alone, this hook is used 3 times.

Almost every invocation of this hook uses different fields in different context, that's a clue for separation.

danny-avila committed:
refactor: decouple connection status from useMCPServerManager with useMCPConnectionStatus

danny-avila committed:
fix: improve MCP tool validation logic to handle unconfigured servers

danny-avila committed:
chore: enhance log message clarity for MCP server disconnection in updateUserPluginsController

danny-avila committed:
refactor: simplify connection status extraction in useMCPConnectionStatus hook

danny-avila committed:
refactor: improve initializing UX

danny-avila committed:
chore: replace string literal with ResourceType constant in useResourcePermissions

danny-avila committed:
refactor: cleanup code, remove redundancies, rename variables for clarity

danny-avila committed:
chore: add back filtering and sorting for mcp tools dialog

danny-avila committed:
refactor: initializeServer to return response and early return

danny-avila committed:
refactor: enhance server initialization logic and improve UI for OAuth interaction

danny-avila committed:
chore: clarify warning message for unconfigured MCP server in handleTools

danny-avila committed:
refactor: prevent CustomUserVarsSection from submitting tools dialog form

danny-avila committed:
fix: nested button of button issue in UninitializedMCPTool

danny-avila committed:
feat: add functionality to revoke custom user variables in MCPToolSelectDialog
